src/pipeline/pipeline.rs

```diff
@@ -124,52 +138,64 @@ impl<S, T, E> Pipeline<S, T>
+
+                    // Track the out body sender. If `self.out_body`
+                    // currently holds a sender for the previous out body, it
+                    // will get dropped. This terminates the stream.
+                    self.out_body = Some(BodySender::Ready(tx));
+
+                    if let Err(_) = self.dispatch.dispatch(Ok(message)) {
+                        // TODO: Should dispatch be infallible
+                        unimplemented!();
+                    }
+                } else {
+                    trace!("read out message");
+
+                    let message = Message::WithoutBody(message);
+
+                    // There is no streaming body. Set `out_body` to `None` so that
+                    // the previous body stream is dropped.
+                    self.out_body = None;
+
+                    if let Err(_) = self.dispatch.dispatch(Ok(message)) {
+                        // TODO: Should dispatch be infalliable
+                        unimplemented!();
+                    }
                 }
             }
-            Frame::MessageWithBody(out_message, body_sender) => {
-                trace!("read out message with body");
-                // Track the out body sender. If `self.out_body`
-                // currently holds a sender for the previous out body, it
-                // will get dropped. This terminates the stream.
-                self.out_body = Some(BodySender::Ready(body_sender));
-
-                if let Err(_) = self.dispatch.dispatch(out_message) {
-                    // TODO: Should dispatch be infallible
-                    unimplemented!();
+            Frame::Body { chunk } => {
+                match chunk {
+                    Some(chunk) => {
+                        trace!("read out body chunk");
+                        try!(self.process_out_body_chunk(chunk));
+                    }
+                    None => {
+                        trace!("read out body EOF");
+                        // Drop the sender.
+                        // TODO: Ensure a sender exists
+                        let _ = self.out_body.take();
+                    }
                 }
             }
-            Frame::Body(Some(chunk)) => {
-                trace!("read out body chunk");
-                try!(self.process_out_body_chunk(chunk));
-            }
-            Frame::Body(None) => {
-                trace!("read out body EOF");
-                // Drop the sender.
-                // TODO: Ensure a sender exists
-                let _ = self.out_body.take();
-            }
             Frame::Done => {
                 trace!("read Frame::Done");
                 // At this point, we just return. This works
                 // because tick() will be called again and go
                 // through the read-cycle again.
                 self.run = false;
             }
-            Frame::Error(_) => {
+            Frame::Error { .. } => {
```

Comment:
Hm is there a way we could perhaps not forget to come back to this? Maybe a panic for now or `unimplemented!`?